### ISSUE TITLE
Gevent integration ignores gevent errors

### ISSUE THREAD

**Issue by wackywendell** [grpc/grpc#15880]
### What version of gRPC and what language are you using?
 
Python 2.7.14, with

```
gevent==1.3.4
grpcio==1.12.1
grpcio-tools==1.8.3
protobuf==3.6.0
```
 
### What operating system (Linux, Windows, …) and version?
 
Saw the same issue on both Linux and OS X.
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
Python 2.7.14
 
### What did you do?

I started a grpc server in go, connected with a Python client using the new gevent integration from #14561, and triggered a `gevent.Timeout`.

See this attached [minimal working example](https://github.com/wackywendell/grpc-gevent/tree/minimal). It includes a GRPC server in Go that sleeps for some time and then returns what it was sent, and a Python client that triggers `gevent.Timeout` before some requests finish.

Also included in the example is a threaded version (using `gevent.threadpool`).
 
### What did you expect to see?

I expected at least the non-threaded version to work smoothly, with no errors, and with contexts canceled correctly for the timed out grpc calls. I was less certain about the threaded version.
 
### What did you see instead?
 
This logged an unusual error:

```
Exception gevent.timeout.Timeout: <Timeout at 0x1058b8be0 seconds=1.5> in 'grpc._cython.cygrpc.run_loop' ignored
```

However, the grpc calls did not timeout, but instead finished, both on the server and client. The exception was ignored, as it said.

### Anything else we should know about your project / environment?

In the minimal working example, it would make more sense to just rely on GRPC timeouts. 

However, in our environment we have a number of different connections we make, many of which are not GRPC, as well as disk I/O and some significant CPU processing.

This means that a timeout set higher up the stack via `gevent.Timeout` may trigger before, after, or during the GRPC calls, and makes more sense for us in addition to a normal GRPC timeout.

#### A "safe" version

Included in my minimal example is a `--safe` option, which triggers the timeout outside the of the `gevent.pool`. With this option, it all works well - all 4 requests are sent, 2 return successfully, the timeout triggers, and the final 2 are canceled on the go side.

I theorize that in this case, the `gevent.timeout` is caught by the outer greenlet which is blocked on the `gevent.pool` code rather than `grpc` code; this outer greenlet then correctly raises the exception. I'm not quite certain how this causes the `grpc` calls to get canceled; perhaps this happens as part of a destructor.

#### `gevent.threadpool`

I also tried combining this code with `gevent.threadpool`, which was quite surprising. This led to ~20k log lines of errors, and no requests received by the go server. I'm not sure if you want to pursue this - supporting `gevent.threadpool` is probably a larger problem with a smaller impact, but it was interesting.

Sprinkled throughout the logging I saw the following error:

```
greenlet.error: cannot switch to a different thread
```

**Comment by wackywendell** [comment]
CCing @kpayson64, as this might be relevant to you.

**Comment by SaurabhBajaj** [comment]
We would love to see some progress on the gevent integration. It's pretty common for most python deployments to have gevent/gunicorn setup. Google libraries like firestore all use grpc and as a result none of those work reliably without the right gevent grpc integration.

**Comment by lidizheng** [comment]
@SaurabhBajaj We are sorry that the gevent integration got de-prioritized. For asynchronous framework, gRPC team now is more focused on `asyncio`. If you observed a gevent integration bug, can you file an issue?

**Comment by wackywendell** [comment]
@lidizheng Does that mean that there is no particular plan to address this issue?

I know prioritization is a complicated thing; I'm just trying to understand your comment and perhaps the status of the original bug on this issue, now that its been open for almost a year.

Thanks! 🙇

**Comment by lidizheng** [comment]
@wackywendell Sorry for the late reply. The gevent support is kind of tricky due to member turnover. I took a closer look at the description. Unfortunately, I could not reproduce the error through your code with gRPC `v1.12` and `v1.21`.

However, I come up with a potential fix in #19356 that should fix the underlying problem.

**Comment by lidizheng** [comment]
@wackywendell @SaurabhBajaj 

Can you test whether this issue is fixed? If so, please close the issue.
You can find our nigntly build at https://packages.grpc.io/

**Comment by wackywendell** [comment]
@lidizheng Thank you for working on this! I appreciate your looking into it.

I'm surprised you couldn't reproduce the error I get; what are you seeing?

I tried again with the updated code you posted (from #19356), and unfortunately, that didn't seem to work for me. I have more details below, but in Python 2 it seems to deadlock, and in Python3 the errors are swallowed and do not surface (and the timeout is ignored).

### Python 3

Using Python3.7.3, gevent==1.4.0, and grpcio@3ab73327, I see errors ignored without any logging whatsoever:

```bash
$ python --version; pip freeze
Python 3.7.3
coverage==4.5.3
Cython==0.29.10
enum34==1.1.6
gevent==1.4.0
greenlet==0.4.15
grpcio==1.22.0.dev0
grpcio-tools==1.8.3
protobuf==3.8.0
six==1.12.0
$ python client.py
2019-06-18 22:21:54.403319 [ 1] Started
2019-06-18 22:21:54.409532 [ 2] Started
2019-06-18 22:21:54.409762 [ 3] Started
2019-06-18 22:21:54.409988 [ 4] Started
2019-06-18 22:21:55.433435 [ 2] Received
2019-06-18 22:21:55.433563 [ 1] Received
2019-06-18 22:21:55.434012 [ 2] Returned
2019-06-18 22:21:55.434090 [ 1] Returned
2019-06-18 22:21:56.431270 [ 4] Received
2019-06-18 22:21:56.431600 [ 4] Returned
2019-06-18 22:21:58.438112 [ 3] Received
2019-06-18 22:21:58.438937 [ 3] Returned
Finished in 0:00:04.043929
```

### Python 2, gevent 1.3.4

With Python2, gevent 1.3.4, it appears to deadlock:

```bash
$ python --version; pip freeze
Python 2.7.16
coverage==4.5.3
Cython==0.29.10
enum34==1.1.6
futures==3.2.0
gevent==1.3.4
greenlet==0.4.15
grpcio==1.22.0.dev0
grpcio-tools==1.21.1
protobuf==3.8.0
six==1.12.0
$ python client.py; date
2019-06-18 22:26:59.048825 [ 1] Started
2019-06-18 22:26:59.049296 [ 2] Started
2019-06-18 22:26:59.049527 [ 3] Started
2019-06-18 22:26:59.049730 [ 4] Started
2019-06-18 22:27:00.068309 [ 1] Received
2019-06-18 22:27:00.069499 [ 2] Received
2019-06-18 22:27:00.070942 [ 1] Returned
2019-06-18 22:27:00.071035 [ 2] Returned
^C^C^C^C^C^C^C^C^C^C^C^C^C^C^C^Z
[1]  + 98869 suspended  python client.py
Tue Jun 18 22:27:09 EDT 2019
```

### Python 2, gevent 1.4.0

```bash
$ pip freeze
coverage==4.5.3
Cython==0.29.10
enum34==1.1.6
futures==3.2.0
gevent==1.4.0
greenlet==0.4.15
grpcio==1.22.0.dev0
grpcio-tools==1.21.1
protobuf==3.8.0
six==1.12.0
$ python client.py; date
2019-06-18 22:29:50.905630 [ 1] Started
2019-06-18 22:29:50.909685 [ 2] Started
2019-06-18 22:29:50.910562 [ 3] Started
2019-06-18 22:29:50.911802 [ 4] Started
2019-06-18 22:29:51.975207 [ 2] Received
2019-06-18 22:29:51.975947 [ 1] Received
2019-06-18 22:29:51.976192 [ 2] Returned
2019-06-18 22:29:51.976235 [ 1] Returned
^C^C^C^C^C^C^C^Z
[1]  + 2709 suspended  python client.py
Tue Jun 18 22:30:18 EDT 2019
```

**Comment by wackywendell** [comment]
@lidizheng I think Github's auto-closer closed this incorrectly. Can you re-open? (I can't - I assume because I'm not a member of `grpc`)

**Comment by lidizheng** [comment]
Thank you for reminding!

**Comment by wackywendell** [comment]
As an update, I have updated my "minimal working example" to use `pipenv` and updated the dependencies, and the issue appears unchanged for these versions:

```
gevent ==1.4.0
grpcio ==1.24.3
grpcio-tools ==1.24.3
protobuf==3.10.0
```

**Comment by apolcyn** [comment]
We acknowledge this to be a valid feature request but since no one has come forward to implement this, we are going to close it for now while maintaining the label/Help wanted to indicate that this can be picked up as future work.

**Comment by beandrad** [comment]
Hi! this seems to have been implemented in [PR 19329](https://github.com/grpc/grpc/pull/19329); would it be possible to get that merged?
